### PR TITLE
PR fixing Windows CI/Travis

### PR DESCRIPTION
**Pull request by StudioEtrange**
I use this PR to check my dev on your Windows CI/Travis environment

So, this is still a work in progress

### CODE REVIEW

**Comment by StudioEtrange** [comment]
I more than half the way
I have a few test more to pass on windows and linux
I will continue tomorrow. It's 2:31 here ;)

* Finally, I implemented the real shell windows command instead of use python os module (except for cd command), to keep your command log logic safe

* I had to review the Config class, and add setter, rework getter and and private members to normalize the path in a easy way

* I disabled color when we are on windows. linux tty code color do not work at all on win :)

**Comment by StudioEtrange** [comment]
For now, on linux CI, I dont understand why these test_shell fails...
```
FAIL gitman/tests/test_shell.py::TestPrograms::()::test_rm_file
FAIL gitman/tests/test_shell.py::TestPrograms::()::test_ln_missing_parent
FAIL gitman/tests/test_shell.py::TestPrograms::()::test_mkdir
FAIL gitman/tests/test_shell.py::TestPrograms::()::test_rm_directory
```

**Comment by jacebrowning** [comment]
> linux tty code color do not work at all on win

This makes it sound like this should work on Windows: https://msdn.microsoft.com/en-us/library/windows/desktop/mt638032(v=vs.85).aspx

What is the error you are seeing? Unless this is causing tests to fail, I'm good implement something like this in another branch: http://stackoverflow.com/a/22254892/429533

> I dont understand why these test_shell fail

In https://github.com/jacebrowning/gitman/commit/7649943be110151272a660eb744f49548103ee93, shell calls were replaced with Python functions. You will want to invert this logic in the tests for commands are are using the shell again.

**Comment by StudioEtrange** [comment]
* About the color on windows : yes it should be possible. On my side I just have tried nothing for now.

* About the revert back of implementing pure windows shell VS using python os module : 
Actually, i found some (rare) case with bugs in python os module on windows. A lot of match on stackoverflow at this subject. So, with this in mind AND to keep clean the fact that you want to output the real commands executed, I manage to implement pure windows shell. So, I remove some of your show(...) function and put call(...) function instead

* About the 4 errors on shell commands on Travis linux, they are all the same. I do not clearly understand what is exactly done with these kind of test : mock_remove.assert_called_once_with('mock/path')
```

> 
> __________________________ TestPrograms.test_rm_file ___________________________
> self = <gitman.tests.test_shell.TestPrograms object at 0x2abff217a470>
> mock_remove = <MagicMock name='remove' id='47003888757728'>
> mock_call = <MagicMock name='call' id='47003888771872'>
>     @patch('os.remove')
>     @patch('os.path.exists', Mock(return_value=True))
>     def test_rm_file(self, mock_remove, mock_call):
>         """Verify the commands to delete files."""
>         shell.rm('mock/path')
> >       mock_remove.assert_called_once_with('mock/path')
> gitman/tests/test_shell.py:73: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
_mock_self = <MagicMock name='remove' id='47003888757728'>
args = ('mock/path',), kwargs = {}
self = <MagicMock name='remove' id='47003888757728'>
msg = "Expected 'remove' to be called once. Called 0 times."
    def assert_called_once_with(_mock_self, *args, **kwargs):
        """assert that the mock was called exactly once and with the specified
            arguments."""
        self = _mock_self
        if not self.call_count == 1:
            msg = ("Expected '%s' to be called once. Called %s times." %
                   (self._mock_name or 'mock', self.call_count))
>           raise AssertionError(msg)
E           AssertionError: Expected 'remove' to be called once. Called 0 times.
/opt/python/3.5.2/lib/python3.5/unittest/mock.py:802: AssertionError
```

**Comment by jacebrowning** [comment]
> implement pure windows shell
> So, is it ok for you ?

Yes, for this tool I'm a fan of a solution that calls out to the shell whenever possible.

**Comment by StudioEtrange** [comment]
* Travis/Linux and Appveyor/Windows tests are all ok

* Except one pep8 error in test_api.py I can not manage to fix it. Every time I change something, I hurt another pep/pylint rule. 

`tests/test_api.py:289:25: E126 continuation line over-indented for hanging indent`

Need help for this

**Comment by StudioEtrange** [comment]
After a lot of commit, I finally managed to pass CI guidelines (pep, pylint)

So, work is done.

Need your review.

File: gitman/commands.py [hunk 1]
```diff
@@ -183,6 +183,7 @@ def delete(*, root=None, force=False):
         common.dedent(level=0)
         common.show("Deleting all dependencies...", color='message', log=False)
         common.show()
+        shell.cd(root)
```
Comment (jacebrowning) [inline gitman/commands.py]: Why is this needed? I'd rather the API functions not need to interact with the shell directly.

File: gitman/models/config.py [hunk 2]
```diff
@@ -1,3 +1,4 @@
+# pylint: disable=attribute-defined-outside-init
```
Comment (jacebrowning) [inline gitman/models/config.py]: I'd rather we not disable this warning and set all attributes in `__init__` (to `None` if needed).

File: gitman/models/config.py [hunk 3]
```diff
@@ -29,22 +30,45 @@ def __init__(self, root, filename="gitman.yml", location="gitman_sources"):
...
     @property
     def config_path(self):
         """Get the full path to the configuration file."""
         return os.path.normpath(os.path.join(self.root, self.filename))
-    path = config_path
+
+    @config_path.setter
+    def config_path(self, value):
+        """Set the full path to the configuration file."""
+        self._config_path = os.path.normpath(value)
```
Comment (jacebrowning) [inline gitman/models/config.py]: When is `self._config_path` used? This setter doesn't seem to be used anywhere.

File: gitman/models/config.py [hunk 4]
```diff
@@ -29,22 +30,45 @@ def __init__(self, root, filename="gitman.yml", location="gitman_sources"):
...
+
+    @property
+    def path(self):
+        return os.path.normpath(self._path)
```
Comment (jacebrowning) [inline gitman/models/config.py]: Calling `normpath` here seems redundant because the setter already called `normpath`.

File: gitman/models/config.py [hunk 5]
```diff
@@ -29,22 +30,45 @@ def __init__(self, root, filename="gitman.yml", location="gitman_sources"):
...
     @property
     def log_path(self):
         """Get the full path to the log file."""
-        return os.path.join(self.location_path, self.LOG)
+        return os.path.normpath(os.path.join(self.location_path, self.LOG))
+
+    @log_path.setter
+    def log_path(self, value):
+        """Set the full path to the log file."""
+        self._log_path = os.path.normpath(value)
```
Comment (jacebrowning) [inline gitman/models/config.py]: Similar comment to above, `self._log_path` doesn't seem to be used anywhere.

File: gitman/models/config.py [hunk 6]
```diff
@@ -29,22 +30,45 @@ def __init__(self, root, filename="gitman.yml", location="gitman_sources"):
...
     def location_path(self):
-        """Get the full path to the dependency storage location."""
-        return os.path.join(self.root, self.location)
+        """Get the full path to the sources location."""
+        return os.path.normpath(os.path.join(self.root, self.location))
+
+    @location_path.setter
+    def location_path(self, value):
+        """Set the full path to the sources location."""
+        self._location_path = os.path.normpath(value)
```
Comment (jacebrowning) [inline gitman/models/config.py]: Similar comment to above, `self._location_path` doesn't seem to be used anywhere.

File: gitman/models/config.py [hunk 7]
```diff
@@ -84,7 +108,11 @@ def install_dependencies(self, *names, depth=None,
                 continue
 
             source.update_files(force=force, fetch=fetch, clean=clean)
-            source.create_link(self.root, force=force)
+            # TODO : windows link support
+            if not os.name == 'nt':
```
Comment (jacebrowning) [inline gitman/models/config.py]: I think that `shell.ln()` is the more appropriate place for this check.

File: gitman/tests/test_shell.py [hunk 8]
```diff
@@ -63,21 +64,25 @@ def test_ln(self, mock_call):
     def test_ln_missing_parent(self, mock_call):
         """Verify the commands to create symbolic links (missing parent)."""
         shell.ln('mock/target', 'mock/source')
-        assert_calls(mock_call, ["ln -s mock/target mock/source"])
+        assert_calls(mock_call, ["mkdir -p mock",
```
Comment (jacebrowning) [inline gitman/tests/test_shell.py]: I would expect this test two have two different results based on `os.name == 'nt`.

File: tests/test_api.py [hunk 9]
```diff
@@ -37,7 +38,7 @@
 
 
 @pytest.fixture
-def config(root="/tmp/gitman-shared"):
+def config(root=os.path.normpath(tempfile.gettempdir() + "/gitman-shared")):
```
Comment (jacebrowning) [inline tests/test_api.py]: `pytest` has a `tmpdir` fixture (http://doc.pytest.org/en/latest/tmpdir.html) that should be used here instead of `tempfile.

Something like:

```
@pytest.fixture
def config(tmpdir, directory="gitman-shared"):
    root = str(tmpdir.join(directory))
    ...
```

File: gitman/commands.py (hunk 1 above)
Comment (StudioEtrange) [inline gitman/commands.py]: * The problem :
On windows we cannot be inside the current directory that we want to remove
I did not manage to find a better place.

* The reasonment :
If I found this cd inside uninstall_dependencies, I will dont know where to cd.

* The solution :
I dont know where to place this elsewhere. So, I will let you find a better, who know better your application architecture, when the PR will be merged

File: gitman/models/config.py (hunk 2 above)
Comment (StudioEtrange) [inline gitman/models/config.py]: * The problem :
I wanted to avoid set to None internal attributes first to None and then just after to the right value

* The solution :
I will re-enable  attribute-defined-outside-init and do something like this

```
self._path = None
self.path = self.config_path

```

File: gitman/models/config.py (hunk 3 above)
Comment (StudioEtrange) [inline gitman/models/config.py]: * The reasonment :
Right. 
It is in prevision of a future usage to not forget.
I prefer near complete POJO-like object to not have to work on it a later time

* The solution :
Do you really want I suppress not yet used setter ?

File: gitman/tests/test_shell.py (hunk 8 above)
Comment (StudioEtrange) [inline gitman/tests/test_shell.py]: But ln is not supported on windows, right ? So this test is not played on windows.
So dont need to write a specific test for windows

File: tests/test_api.py (hunk 9 above)
Comment (StudioEtrange) [inline tests/test_api.py]: If you are ok, I will let you do this, when the PR is ready. I dont know enough how pytest and fixture work.

File: gitman/models/config.py (hunk 4 above)
Comment (StudioEtrange) [inline gitman/models/config.py]: Right.

Comment (StudioEtrange) [inline gitman/models/config.py]: * Done

File: gitman/models/config.py (hunk 2 above)
Comment (StudioEtrange) [inline gitman/models/config.py]: * Done

File: gitman/models/config.py (hunk 5 above)
Comment (StudioEtrange) [inline gitman/models/config.py]: see answer for self._config_pat

File: gitman/models/config.py (hunk 6 above)
Comment (StudioEtrange) [inline gitman/models/config.py]: see answer for self._config_path

File: gitman/models/config.py (hunk 3 above)
Comment (jacebrowning) [inline gitman/models/config.py]: Yes, I'd prefer not to add code that is unused. These properties should be read-only.

**Comment by StudioEtrange** [comment]
Done, for now.

**Comment by jacebrowning** [comment]
@StudioEtrange There's a lot of great stuff on this PR, but it's getting kind of large. If it's OK with you, I'll like to manually pull in the different pieces. Thanks for helping drive a better architecture!

**Comment by StudioEtrange** [comment]
@jacebrowning As you wish. I will stop here anyway.
I think it will be pretty hard to pick piece by piece, but you can try :)


For now :
* All the unit tests are ok on Linux/Windows. 
* And all the Pylint stuff are ok

For later :
* Find a way to support windows shell color
* Support symlink on windows
* Improve makefile



Now, I am done :)

**Comment by StudioEtrange** [comment]
@jacebrowning 
Do you know why I can launch pytest only for gitman folder but not on tests folder ?
I tried `pytest gitman tests`. But unit test in `tests `folder are never launched on my windows workstation.
So I permanently have to use your Appveyor/Travis infrastructure.

**Comment by StudioEtrange** [comment]
@jacebrowning 
Ok you pick code piece by piece. And try to fix some things or arrange some things at the same time.
I was thinking you will pick several bag of commits as is.

Are you really sure you dont want juste merge the PR, and fix after that the one or two things that you dislike ? There is a lot of chance that you will re-do the bug I have fixed while I was working on this PR and windows support.

Just merge the PR, and change things after, really. :)

**Comment by jacebrowning** [comment]
@StudioEtrange Try running:

`$ TEST_INTEGRATION=true pytest gitman tests`

I try to keep unit tests separate from integration tests, but in this project, some of the integration tests are mixed.

**Comment by jacebrowning** [comment]
@StudioEtrange I have incorporated these changes into `GitMan==1.2b1`. Please test.

**Comment by StudioEtrange** [comment]
@jacebrowning 
Do you mean : test the branch release/v1.2 ?
I cant find GitMan==1.2b1 anywhere

**Comment by jacebrowning** [comment]
@StudioEtrange `1.2b1` is a beta release on PyPI, but testing the release branch works too.

**Comment by jacebrowning** [comment]
I'll release this version is you say it works for you on Windows.

**Comment by StudioEtrange** [comment]
I have tested the one in branch release/v1.2
on Windows

It seems to be ok !

**Comment by jacebrowning** [comment]
Thanks! All changes are now part of `GitMan==1.2`.
